### PR TITLE
NativeRegExp: Implement named capture groups

### PR DESCRIPTION
**Pull request by balajirrao**
**Note: This PR also includes commits from the lookbehind [PR](https://github.com/mozilla/rhino/pull/1843).**

This adds partial support for named capture groups in NativeRegExp. This also adds support for [duplicate-named-capturing-groups](https://github.com/tc39/proposal-duplicate-named-capturing-groups).

What is not implemented yet:
- capture group names that consist of unicode escape sequences.
- String.prototype.{replace, replaceAll}
 
Many new test262 test cases pass and there are no regressions. The ones that are failing are due to
- 'u' regexp flag not supported
- objects will null prototype do not work correctly with respect to accessing `__proto__` (a fix is being worked on)
- existing bugs in String.prototype.{replace, replaceAll, split}
- no support for classes

### CODE REVIEW

**Comment by rbri** [comment]
Great

**Comment by rbri** [comment]
@balajirrao thanks for all the work on the regexp support.

Since ages HtmlUnit uses the rhino regexp more as a backup and tries to translate the js regexp into java regexp. Based on your work i like to remove that stuff and stay with the rhino impl.
But while updating i found another thing - web pages also using the js regexp support for validating input fields. This means i have to use the regexp stuff without a js environment (more or less). Or long story short...

What do you think about separating the regexp stuff from the js stuff meaning moving most of the static methods and all the helper classes into a separate (static) class?
Have done some prototyping and it seems to work well. Only some methods still use a Context parameter. Currently i think about options for that.
This might also help a bit to write more basic unit tests.

So what do you think?

**Comment by balajirrao** [comment]
@rbri thanks for your comment!

What you are saying sounds very reasonable. I have a couple of regexp improvements on the plate before I can tackle this. In case you have your prototype on a branch, I'd be interested to take a look!

**Comment by rbri** [comment]
@ balajirrao

> I have a couple of regexp improvements in the pipeline before I can tackle this.
Yes please....

> In case you have your prototype on a branch, I'd be interested to take a look!
I did something in this direction but i was not really happy with my solution.

Will work on that maybe during the next days and prepare a draft pr to have some base to discuss.

**Comment by rbri** [comment]
@balajirrao 

As starting point.... 56567bc71c4150285a44b3e60857fb05ad7dd696

**Comment by balajirrao** [comment]
@gbrail I figured you must have started reviewing this and hence I've pushed some changes I'd like to be included in the PR, in a separate commit. The changes improves how named capture groups are parsed and has some refactoring to pave the way for unicode support (which I'm working on ATM). The commit message has some info. 

Some background:
While working on unicode regexp support , I discovered https://tc39.es/ecma262/#sec-regular-expressions-patterns, which I also figured is what Rhino supports. This led me to extract the parser for the CharacterClassEscape and CharacterEscape non-terminals,  which can be used in parsing both the AtomEscape and the ClassEscape non-terminals (the code to do this was duplicated in 3 different places in the code before).

File: rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java [hunk 1]
```diff
@@ -888,150 +884,46 @@ private static boolean parseAlternative(CompilerState state) {
...
             }
-            if ((state.flags & JSREG_FOLD) != 0) {
-                char cu = upcase((char) localMax);
-                char cd = downcase((char) localMax);
-                localMax = (cu >= cd) ? cu : cd;
+        }
+
+        for (RENode node : classContents.escapeNodes) {
+            if (node.op != REOP_FLAT) {
+                target.bmsize = 65536;
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:921]: Can we use a named constant instead of a magic number?

File: rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java [hunk 2]
```diff
@@ -1175,7 +1067,285 @@ private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend
...
+            ranges = new ArrayList<>();
+            escapeNodes = new ArrayList<>();
+            sense = true;
+        }
+    }
+
+    private static ClassContents parseClassContents(CompilerState state, ParserParameters params) {
+        char[] src = state.cpbegin;
+        char rangeStart = 0;
+        ;
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:1282]: Spurious

File: rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java [hunk 3]
```diff
@@ -1175,7 +1067,285 @@ private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend
...
+        ArrayList<Character> ranges; // ranges stored as (start1, end1, start2, end2, ...)
+        ArrayList<RENode> escapeNodes;
+
+        ClassContents() {
+            chars = new ArrayList<>();
+            ranges = new ArrayList<>();
+            escapeNodes = new ArrayList<>();
+            sense = true;
+        }
+    }
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:1277]: Why not do this in the field declaration?

```suggestion
    static class ClassContents {
        boolean sense = true;
        ArrayList<Character> chars = new ArrayList<>();
        ArrayList<Character> ranges = new ArrayList<(); // ranges stored as (start1, end1, start2, end2, ...)
        ArrayList<RENode> escapeNodes = new ArrayList<>();
    }
```

**Review by gbrail — CHANGES_REQUESTED** [review]
Thanks -- great progress, it looks like we have some minor comments from me and also Andrea so please take a look.

Also, I didn't catch this in a previous, but I see that the class now has a "debug" flag. We actually have a whole mechanism for those that lets you easily set it at runtime -- take a look at what's done in the class below and please do it the same way:

https://github.com/mozilla/rhino/blob/f8ce76a38b172521a9e49050ebfb83de0bb5e073/rhino/src/main/java/org/mozilla/javascript/Token.java#L29

Finally, please take a look at test coverage for the new code -- I didn't look yet. Thanks!

File: rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java [hunk 4]
```diff
@@ -78,7 +82,8 @@ public class NativeRegExp extends IdScriptableObject {
     //    private static final byte REOP_UCFLATi       = 21; /* case-independent REOP_UCFLAT */
     private static final byte REOP_CLASS = 22; /* character class with index */
     private static final byte REOP_NCLASS = 23; /* negated character class with index */
-    private static final byte REOP_SIMPLE_END = 23; /* end of 'simple opcodes' */
+    private static final byte REOP_NAMED_BACKREF = 24; /* named back-reference */
+    private static final byte REOP_SIMPLE_END = 24; /* end of 'simple opcodes' */
```
[the remark this answers is not in the document]
Comment (balajirrao) [reply rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:86]: It is fine for them to have the same values. REOP_SIMPLE_END is not an operator, but used as a marker to determine if an op is simple using `op <= REOP_SIMPLE_END`. 

The `+1` pattern is good. I'll switch them in a next PR, right after this one :)

File: rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java [hunk 5]
```diff
@@ -536,6 +550,61 @@ private static void prettyPrintRE(RECompiled regexp) {
...
+    private static void extractNamedCaptureGroups(
+            char[] src, RENode re, Map<String, List<Integer>> namedCaptureGroups) {
+        RENode node = re;
+        while (node != null) {
+            if (node.op == REOP_LPAREN) {
+                if (node.captureGroupNameLength != 0) {
+                    String name =
+                            new String(
+                                    src, node.captureGroupNameIndex, node.captureGroupNameLength);
+                    ArrayList<Integer> entry = new ArrayList<>(1);
```
[the remark this answers is not in the document]
Comment (balajirrao) [reply rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:562]: Not quite. I create an arraylist of size 1 optimistically assuming that there are no duplicates. However, if we find duplicates (allowed in the case of disjunctions), the list will have more elements than one. I think this deserves a comment in the code.

File: rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java [hunk 6]
```diff
@@ -1061,7 +1035,317 @@ private static RENode reverseNodeList(RENode head) {
         return prev;
     }
 
-    private static boolean parseTerm(CompilerState state) {
+    private static String extractCaptureGroupName(char[] src, int cpbegin, int cpend) {
+        // TODO: support capture group names consisting of unicode escape sequences
+        String srcString = new String(src);
```
[the remark this answers is not in the document]
Comment (balajirrao) [reply rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:1040]: Good catch - I'm not very familiar with the unicode api in Character and String. As soon I as I got something working I was happy to move on 😅. Will fix.

**Comment by balajirrao** [comment]
@gbrail thank you for taking a look! I've addressed the comments. I also took at look at the test coverage and added couple tests for a couple of edge cases.

I'll fix the debug flag in a next PR - it has been around before I started with Rhino.

**Comment by gbrail** [comment]
Can you please just comment that constant with a duplicate value? Other
people will look at it and assume it's a mistake.

**Comment by gbrail** [comment]
Thanks for all your work on this!

**Comment by balajirrao** [comment]
@gbrail thank you for your review and for merging it! I think I'm finally getting the hang of the regexp parser.
